### PR TITLE
ARROW-14169: [R] altrep for factors

### PR DESCRIPTION
**Pull request by romainfrancois**
replaces #11402

### CODE REVIEW

File: r/src/altrep.cpp [hunk 1]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+      }
+    }
+
+    return slice->length();
+  }
+
+  template <typename Transpose>
+  static void GetRegionDispatch(const std::shared_ptr<Array>& array,
+                                const std::shared_ptr<Array>& indices,
+                                Transpose transpose, int* out) {
```
Comment (pitrou) [inline r/src/altrep.cpp:603]: Pass `Transpose&& transpose` and pass `std::forward<Transpose>(transpose)` below.

File: r/src/altrep.cpp [hunk 2]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+        GetRegionTranspose<UInt16Type>(array, indices, transpose, out);
+        break;
+      case Type::INT16:
+        GetRegionTranspose<Int16Type>(array, indices, transpose, out);
+        break;
+      case Type::INT32:
+        GetRegionTranspose<Int32Type>(array, indices, transpose, out);
+        break;
+      case Type::UINT32:
+        GetRegionTranspose<Int32Type>(array, indices, transpose, out);
```
Comment (pitrou) [inline r/src/altrep.cpp:621]: `UInt32Type`

File: r/src/altrep.cpp [hunk 3]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+        break;
+      default:
```
Comment (pitrou) [inline r/src/altrep.cpp:623]: Why are 64-bit indices not handled?

File: r/src/altrep.cpp [hunk 4]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+        break;
+    }
+  }
+
+  template <typename Type, typename Transpose>
+  static void GetRegionTranspose(const std::shared_ptr<Array>& array,
+                                 const std::shared_ptr<Array>& indices,
+                                 Transpose transpose, int* out) {
+    using index_type = typename arrow::TypeTraits<Type>::ArrayType::value_type;
```
Comment (pitrou) [inline r/src/altrep.cpp:632]: `using index_type = typename Type::c_type` should work.

File: r/src/altrep.cpp [hunk 5]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+    auto raw_indices = indices->data()->GetValues<index_type>(1);
+
+    auto n = array->length();
+
+    // then set the R NA sentinels if needed
+    if (indices->null_count() > 0) {
```
Comment (pitrou) [inline r/src/altrep.cpp:638]: Can use `VisitArrayDataInline` here:
```c++
VisitArrayDataInline<Type>(*array->data(),
  /*valid_func=*/ [&](index_type index) {
    *out++ = transpose(index) + 1;
  },
  /*null_func=*/ [&]() {
    *out++ = cpp11::na<int>();
  });
```

File: r/src/altrep.cpp [hunk 6]
```diff
@@ -366,6 +388,277 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
+
+  // The CADR(data2) is used to store the transposed arrays when unification is needed
+  // In that case we store a vector of Buffers
+  using BufferVector = std::vector<std::shared_ptr<Buffer>>;
+
+  static bool WasUnified(SEXP alt) { return !Rf_isNull(CADR(R_altrep_data2(alt))); }
+
+  static const std::shared_ptr<Buffer>& GetArrayTransposed(SEXP alt, int i) {
+    const auto& arrays = *Pointer<BufferVector>(CADR(R_altrep_data2(alt)));
+    return arrays->operator[](i);
```
Comment (pitrou) [inline r/src/altrep.cpp:412]: Nit: can write `(*arrays)[i]`.

**Review by pitrou — CHANGES_REQUESTED** [review]
Some comments below.

**Comment by jonkeane** [comment]
@ursabot please benchmark

File: r/src/altrep.cpp (hunk 5 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:638]: Nice. We can probably use this in other places.

**Comment by nealrichardson** [comment]
@jonkeane are we trying to land this for 7.0.0?

**Comment by jonkeane** [comment]
It would be great if we could — @pitrou are you satisfied with the changes you requested above? 

It would be worthwhile to rebase + rebenchmark this if we can, but I can review the R side as well.

**Review by pitrou — COMMENTED** [review]
Just one more thing, otherwise LGTM.

**Comment by jonkeane** [comment]
@ursabot please benchmark

**Comment by jonkeane** [comment]
I rebased this to benchmark it again with a more recent branching point — but didn't change anything else

**Comment by jonkeane** [comment]
Have you had a chance to look at this @romainfrancois ?

**Comment by romainfrancois** [comment]
I haven't unfortunately. Will try on Monday morning.

**Review by jonkeane — APPROVED** [review]

**Pull request closed without merging**